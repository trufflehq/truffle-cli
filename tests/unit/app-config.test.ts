import { describe, expect, it } from 'vitest';
import {
  convertActionConfigsToMothertreeActionConfigs,
  convertAppConfigToMothertreeConfig,
  convertProductConfigsToMothertreeProductAndVariantConfigs,
  validateAppConfig,
} from '../../src/util/app-config';

describe('app-config', () => {
  describe('validateAppConfig', () => {
    it('should allow a valid config', () => {
      const config = {
        path: '@truffle/test-app',
        name: 'test-app',
        cliVersion: '0.0.0',
        embeds: [
          {
            slug: 'test-embed',
            url: 'https://example.com',
            contentPageType: 'youtubeVideo',
          },
        ],
        countables: [
          {
            slug: 'channel-points',
            name: 'Channel Points',
            category: 'currency-virtual',
            decimalPlaces: 0,
          },
        ],
        products: [
          {
            slug: 'test-product',
            name: 'Test Product',
            category: 'test-category',
            variants: [
              {
                slug: 'test-variant',
                price: 100,
              },
            ],
          },
        ],
      };

      expect(validateAppConfig(config)).toMatchSnapshot();
    });

    it('should throw an error for an invalid config', () => {
      const config = {
        path: '@truffle/test-app',
        name: 'test-app',
        // missing cliVersion
      };

      expect(() => validateAppConfig(config)).toThrowErrorMatchingSnapshot();
    });
  });

  describe('convertActionConfigsToMothertreeActionConfigs', () => {
    it('should convert a basic webhook action', () => {
      const actionConfigs = [
        {
          slug: 'test-action',
          operation: 'webhook',
          url: 'https://example.com',
        },
      ];
      const mtAppConfig = {
        actions: [],
      };

      convertActionConfigsToMothertreeActionConfigs(actionConfigs, mtAppConfig);

      expect(mtAppConfig.actions).toMatchSnapshot();
    });

    it('should convert a basic webhook action with data', () => {
      const actionConfigs = [
        {
          slug: 'test-action',
          operation: 'webhook',
          url: 'https://example.com',
          data: {
            some: 'data',
          },
        },
      ];
      const mtAppConfig = {
        actions: [],
      };

      convertActionConfigsToMothertreeActionConfigs(actionConfigs, mtAppConfig);

      expect(mtAppConfig.actions).toMatchSnapshot();
    });

    it('should convert a basic workflow action', () => {
      const actionConfigs = [
        {
          slug: 'test-action',
          operation: 'workflow',
          strategy: 'sequential',
          actions: [
            './_Action/action-1',
            {
              slug: 'action-2',
              operation: 'webhook',
              url: 'https://example.com',
            },
          ],
        },
      ];
      const mtAppConfig = {
        actions: [],
      };

      convertActionConfigsToMothertreeActionConfigs(actionConfigs, mtAppConfig);

      expect(mtAppConfig.actions).toMatchSnapshot();
    });

    it('should convert a basic exchange action', () => {
      const actionConfigs = [
        {
          slug: 'test-action',
          operation: 'exchange',
          assets: [
            {
              path: './_Asset/test-asset',
              quantity: 1,
            },
          ],
        },
      ];
      const mtAppConfig = {
        actions: [],
      };

      convertActionConfigsToMothertreeActionConfigs(actionConfigs, mtAppConfig);

      expect(mtAppConfig.actions).toMatchSnapshot();
    });

    it('should convert a basic apply-powerup action with a powerup path', () => {
      const actionConfigs = [
        {
          slug: 'test-action',
          operation: 'apply-powerup',
          powerup: '@truffle/app/_Powerup/test-powerup',
          targetType: 'test-target-type',
          targetId: 'test-target-id',
          ttlSeconds: 60,
        },
      ];
      const mtAppConfig = {
        actions: [],
      };

      convertActionConfigsToMothertreeActionConfigs(actionConfigs, mtAppConfig);

      expect(mtAppConfig.actions).toMatchSnapshot();
    });

    it('should convert a basic apply-powerup action with a powerup object', () => {
      const actionConfigs = [
        {
          slug: 'test-action',
          operation: 'apply-powerup',
          powerup: {
            slug: 'test-powerup',
            name: 'Test Powerup',
          },
          targetType: 'test-target-type',
          targetId: 'test-target-id',
          ttlSeconds: 60,
        },
      ];
      const mtAppConfig = {
        powerups: [],
        actions: [],
      };

      convertActionConfigsToMothertreeActionConfigs(actionConfigs, mtAppConfig);

      expect(mtAppConfig).toMatchSnapshot();
    });
  });

  describe('convertProductConfigsToMothertreeProductAndVariantConfigs', () => {
    it('should convert a basic product', () => {
      const productConfigs = [
        {
          slug: 'test-product',
          name: 'Test Product',
          category: 'test-category',
          variants: [
            {
              slug: 'test-variant',
              price: 100,
            },
          ],
        },
      ];
      const mtAppConfig = {
        products: [],
        productVariants: [],
        actions: [],
      };

      convertProductConfigsToMothertreeProductAndVariantConfigs(
        productConfigs,
        mtAppConfig,
      );

      expect(mtAppConfig).toMatchSnapshot();
    });

    it('should convert a product with a variant with an action', () => {
      const productConfigs = [
        {
          slug: 'test-product',
          name: 'Test Product',
          category: 'test-category',
          variants: [
            {
              slug: 'test-variant',
              price: 100,
              action: {
                slug: 'test-action',
                operation: 'webhook',
                url: 'https://example.com',
              },
            },
          ],
        },
      ];
      const mtAppConfig = {
        products: [],
        productVariants: [],
        actions: [],
      };

      convertProductConfigsToMothertreeProductAndVariantConfigs(
        productConfigs,
        mtAppConfig,
      );

      expect(mtAppConfig).toMatchSnapshot();
    });
  });

  describe('convertAppConfigToMothertreeConfig', () => {
    it('should convert a basic app config', () => {
      const appConfig = {
        path: '@truffle/test-app',
        name: 'test-app',
        cliVersion: '0.0.0',
        embeds: [
          {
            slug: 'test-embed',
            url: 'https://example.com',
            contentPageType: 'youtubeVideo',
          },
        ],
        actions: [
          {
            slug: 'test-action',
            operation: 'webhook',
            url: 'https://example.com',
          },
        ],
        countables: [
          {
            slug: 'channel-points',
            name: 'Channel Points',
            category: 'currency-virtual',
            decimalPlaces: 0,
          },
        ],
        products: [
          {
            slug: 'test-product',
            name: 'Test Product',
            category: 'test-category',
            variants: [
              {
                slug: 'test-variant',
                price: 100,
              },
            ],
          },
        ],
      };

      expect(convertAppConfigToMothertreeConfig(appConfig)).toMatchSnapshot();
    });

    it('should define a postInstallActionPath for a postInstallAction that is defined as a path', () => {
      const appConfig = {
        path: '@truffle/test-app',
        name: 'test-app',
        cliVersion: '0.0.0',
        postInstallAction: './_Action/post-install',
      };

      expect(
        convertAppConfigToMothertreeConfig(appConfig).postInstallActionPath,
      ).toBe(appConfig.postInstallAction);
    });

    it('should convert a postInstallAction that is defined as an action object', () => {
      const appConfig = {
        path: '@truffle/test-app',
        name: 'test-app',
        cliVersion: '0.0.0',
        postInstallAction: {
          operation: 'webhook',
          url: 'https://example.com',
        },
      };

      expect(convertAppConfigToMothertreeConfig(appConfig)).toMatchSnapshot();
    });
  });
});
