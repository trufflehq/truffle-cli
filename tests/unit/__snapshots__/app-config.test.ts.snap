// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`app-config > convertActionConfigsToMothertreeActionConfigs > should convert a basic apply-powerup action with a powerup object 1`] = `
{
  "actions": [
    {
      "inputsTemplate": {
        "powerupPath": "./_Powerup/test-powerup",
        "targetId": "test-target-id",
        "targetType": "test-target-type",
        "ttlSeconds": 60,
      },
      "isDirectExecutionAllowed": false,
      "operation": "apply-powerup",
      "slug": "test-action",
    },
  ],
  "powerups": [
    {
      "name": "Test Powerup",
      "slug": "test-powerup",
    },
  ],
}
`;

exports[`app-config > convertActionConfigsToMothertreeActionConfigs > should convert a basic apply-powerup action with a powerup path 1`] = `
[
  {
    "inputsTemplate": {
      "powerupPath": "@truffle/app/_Powerup/test-powerup",
      "targetId": "test-target-id",
      "targetType": "test-target-type",
      "ttlSeconds": 60,
    },
    "isDirectExecutionAllowed": false,
    "operation": "apply-powerup",
    "slug": "test-action",
  },
]
`;

exports[`app-config > convertActionConfigsToMothertreeActionConfigs > should convert a basic exchange action 1`] = `
[
  {
    "inputsTemplate": {
      "assets": [
        {
          "path": "./_Asset/test-asset",
          "quantity": 1,
        },
      ],
    },
    "isDirectExecutionAllowed": false,
    "operation": "exchange",
    "slug": "test-action",
  },
]
`;

exports[`app-config > convertActionConfigsToMothertreeActionConfigs > should convert a basic webhook action 1`] = `
[
  {
    "inputsTemplate": {
      "data": undefined,
      "url": "https://example.com",
    },
    "isDirectExecutionAllowed": false,
    "operation": "webhook",
    "slug": "test-action",
  },
]
`;

exports[`app-config > convertActionConfigsToMothertreeActionConfigs > should convert a basic webhook action with data 1`] = `
[
  {
    "inputsTemplate": {
      "data": {
        "some": "data",
      },
      "url": "https://example.com",
    },
    "isDirectExecutionAllowed": false,
    "operation": "webhook",
    "slug": "test-action",
  },
]
`;

exports[`app-config > convertActionConfigsToMothertreeActionConfigs > should convert a basic workflow action 1`] = `
[
  {
    "inputsTemplate": {
      "data": undefined,
      "url": "https://example.com",
    },
    "isDirectExecutionAllowed": false,
    "operation": "webhook",
    "slug": "test-action-step-1",
  },
  {
    "inputsTemplate": {
      "actionPaths": [
        "./_Action/action-1",
        "./_Action/test-action-step-1",
      ],
      "strategy": "sequential",
    },
    "isDirectExecutionAllowed": false,
    "operation": "workflow",
    "slug": "test-action",
  },
]
`;

exports[`app-config > convertAppConfigToMothertreeConfig > should convert a basic app config 1`] = `
{
  "actions": [
    {
      "inputsTemplate": {
        "data": undefined,
        "url": "https://example.com",
      },
      "isDirectExecutionAllowed": false,
      "operation": "webhook",
      "slug": "test-action",
    },
    {
      "inputsTemplate": {
        "assets": "{{USE_SECURE_PROVIDED}}",
      },
      "operation": "exchange",
      "slug": "default-exchange-action",
    },
  ],
  "cliVersion": "0.0.0",
  "countables": [
    {
      "category": "currency-virtual",
      "decimalPlaces": 0,
      "name": "Channel Points",
      "slug": "channel-points",
    },
  ],
  "embeds": [
    {
      "contentPageType": "youtubeVideo",
      "slug": "test-embed",
      "url": "https://example.com",
    },
  ],
  "name": "test-app",
  "path": "@truffle/test-app",
  "powerups": [],
  "productVariants": [
    {
      "actionPath": "./_Action/default-exchange-action",
      "assetTemplates": [
        {
          "count": 100,
          "entityPath": "@truffle/sparks/_Countable/sparks:@truffle",
          "metadata": {},
          "receivers": [
            {
              "entityId": "{{USE_ORG_ID}}",
              "entityType": "org",
              "share": 1,
            },
          ],
          "senders": [
            {
              "entityId": "{{USE_USER_ID}}",
              "entityType": "user",
              "share": 1,
            },
          ],
        },
      ],
      "description": undefined,
      "name": undefined,
      "productPath": "./_Product/test-product",
      "slug": "test-variant",
    },
  ],
  "products": [
    {
      "category": "test-category",
      "name": "Test Product",
      "rank": 0,
      "slug": "test-product",
    },
  ],
}
`;

exports[`app-config > convertAppConfigToMothertreeConfig > should convert a postInstallAction that is defined as an action object 1`] = `
{
  "actions": [
    {
      "inputsTemplate": {
        "data": undefined,
        "url": "https://example.com",
      },
      "isDirectExecutionAllowed": false,
      "operation": "webhook",
      "slug": "post-install-action",
    },
  ],
  "cliVersion": "0.0.0",
  "countables": [],
  "embeds": [],
  "name": "test-app",
  "path": "@truffle/test-app",
  "postInstallActionPath": "./_Action/post-install-action",
  "powerups": [],
  "productVariants": [],
  "products": [],
}
`;

exports[`app-config > convertProductConfigsToMothertreeProductAndVariantConfigs > should convert a basic product 1`] = `
{
  "actions": [
    {
      "inputsTemplate": {
        "assets": "{{USE_SECURE_PROVIDED}}",
      },
      "operation": "exchange",
      "slug": "default-exchange-action",
    },
  ],
  "productVariants": [
    {
      "actionPath": "./_Action/default-exchange-action",
      "assetTemplates": [
        {
          "count": 100,
          "entityPath": "@truffle/sparks/_Countable/sparks:@truffle",
          "metadata": {},
          "receivers": [
            {
              "entityId": "{{USE_ORG_ID}}",
              "entityType": "org",
              "share": 1,
            },
          ],
          "senders": [
            {
              "entityId": "{{USE_USER_ID}}",
              "entityType": "user",
              "share": 1,
            },
          ],
        },
      ],
      "description": undefined,
      "name": undefined,
      "productPath": "./_Product/test-product",
      "slug": "test-variant",
    },
  ],
  "products": [
    {
      "category": "test-category",
      "name": "Test Product",
      "rank": 0,
      "slug": "test-product",
    },
  ],
}
`;

exports[`app-config > convertProductConfigsToMothertreeProductAndVariantConfigs > should convert a product with a variant with an action 1`] = `
{
  "actions": [
    {
      "inputsTemplate": {
        "assets": "{{USE_SECURE_PROVIDED}}",
      },
      "operation": "exchange",
      "slug": "default-exchange-action",
    },
    {
      "inputsTemplate": {
        "data": undefined,
        "url": "https://example.com",
      },
      "isDirectExecutionAllowed": false,
      "operation": "webhook",
      "slug": "test-product-test-variant-workflow-step-1",
    },
    {
      "inputsTemplate": {
        "actionPaths": [
          "./_Action/default-exchange-action",
          "./_Action/test-product-test-variant-workflow-step-1",
        ],
        "strategy": "sequential",
      },
      "isDirectExecutionAllowed": false,
      "operation": "workflow",
      "slug": "test-product-test-variant-workflow",
    },
  ],
  "productVariants": [
    {
      "actionPath": "./_Action/test-product-test-variant-workflow",
      "assetTemplates": [
        {
          "count": 100,
          "entityPath": "@truffle/sparks/_Countable/sparks:@truffle",
          "metadata": {},
          "receivers": [
            {
              "entityId": "{{USE_ORG_ID}}",
              "entityType": "org",
              "share": 1,
            },
          ],
          "senders": [
            {
              "entityId": "{{USE_USER_ID}}",
              "entityType": "user",
              "share": 1,
            },
          ],
        },
      ],
      "description": undefined,
      "name": undefined,
      "productPath": "./_Product/test-product",
      "slug": "test-variant",
    },
  ],
  "products": [
    {
      "category": "test-category",
      "name": "Test Product",
      "rank": 0,
      "slug": "test-product",
    },
  ],
}
`;

exports[`app-config > validateAppConfig > should allow a valid config 1`] = `
{
  "cliVersion": "0.0.0",
  "countables": [
    {
      "category": "currency-virtual",
      "decimalPlaces": 0,
      "name": "Channel Points",
      "slug": "channel-points",
    },
  ],
  "embeds": [
    {
      "contentPageType": "youtubeVideo",
      "slug": "test-embed",
      "url": "https://example.com",
    },
  ],
  "name": "test-app",
  "path": "@truffle/test-app",
  "products": [
    {
      "category": "test-category",
      "name": "Test Product",
      "slug": "test-product",
      "variants": [
        {
          "price": 100,
          "slug": "test-variant",
        },
      ],
    },
  ],
}
`;

exports[`app-config > validateAppConfig > should throw an error for an invalid config 1`] = `
[ValidationError: {
  "path": "@truffle/test-app",
  "name": "test-app",
  [41m"cliVersion"[0m[31m [1]: -- missing --[0m
}
[31m
[1] "cliVersion" is required[0m]
`;
